### DISCUSSION TITLE
Query compilation cache metrics

### DISCUSSION THREAD

**Discussion by tonycosentini** [sqlalchemy/sqlalchemy#10722] [Ideas]
Hi,

Sorry if there is already an API for this, but I was wondering if it would be possible to add some kind of way to track stats for the query compilation cache on an engine.

In particular, I've run into issues similar to [this one](https://github.com/sqlalchemy/sqlalchemy/discussions/9771) where a dynamic `prefix_with` causes lots of cache thrashing.

It would be great if there was a mechanism to track the oldest entry in the LRU cache to ensure it's not constantly getting rotated.

**Comment by CaselIT** [comment]
Hi,

The internal cache currently is not settable when constructing an engine, but you could pass a mapping to `execution_options`.
The internal mapping used by the cache is `util.LRUCache` (that's not really private, not really public) that takes a callable that's called when the cache is pruning items.

Something like this may work
```py
orig_engine = create_engine(...)
lru = util.LRUCache(limit_here, size_alert=my_notify_when_cache_is_pruning)
engine = orig_engine.execution_options(compiled_cache=lru)
```

**Comment by zzzeek** [comment]
the general stats are reported in logging which is described at https://docs.sqlalchemy.org/en/20/core/connections.html#estimating-cache-performance-using-logging 

you can use the logging alone to detect cache thrashing, if you see lots of queries being "generated" over and over.

if you wanted to peek into the LRU cache that's pretty doable, it's there at `engine._compiled_cache` and it uses [LRUCache](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/util/_collections.py#L484); however I would mostly do @CaselIT 's suggestion above, set an LRUCache with a size alert.  when the cache is pruning, that means it filled up, which suggests too many unique objects are being put into it.

**Comment by tonycosentini** [comment]
I think both comments provided solutions that tackle what I'm trying to do. Thanks @CaselIT + @zzzeek! (and sorry for the crazy long delay)
